### PR TITLE
[FEAT] chain proxy

### PR DESCRIPTION
**Pull request by gabrielbazan7**

### CODE REVIEW

**Comment by micahriggan** [comment]
Nice work on this, I believe you can consolidate BCH into this https://github.com/bitpay/bitcore/blob/768dff7b9717a864e79ea34f6f025059bf9c819b/packages/bitcore-wallet-service/src/lib/chain/bch/index.ts

File: packages/bitcore-wallet-service/src/lib/chain/bch/index.ts [hunk 1]
```diff
@@ -0,0 +1,100 @@
+import { BitcoreLibCash } from 'crypto-wallet-core';
```
Comment (micahriggan) [inline packages/bitcore-wallet-service/src/lib/chain/bch/index.ts]: No types on anything :(

File: packages/bitcore-wallet-service/src/lib/server.ts [hunk 2]
```diff
@@ -1778,67 +1779,7 @@ export class WalletService {
...
-                byAddress[utxo.address].amount += utxo.satoshis;
-              });
-
-              balance.byAddress = _.values(byAddress);
-
-              return cb(null, balance);
-            }
-          );
-        }
+        return ChainService.getWalletBalance(opts, cb);
```
Comment (micahriggan) [inline packages/bitcore-wallet-service/src/lib/server.ts]: not passing wallet in here, but the method loads from wallet.coin

```
return ChainService.getWalletBalance(opts, cb);
```

Types would have caught this one

File: packages/bitcore-wallet-service/src/lib/server.ts [hunk 3]
```diff
@@ -2879,52 +2636,16 @@ export class WalletService {
...
-
-                          gasLimit = inGasLimit || Defaults.DEFAULT_GAS_LIMIT;
-                          opts.fee = feePerKb * gasLimit;
-                          return next();
-                        });
-                    } else {
-                      next();
-                    }
-                  });
+                  feePerKb = await ChainService.getFeePerKb(wallet, opts);
```
Comment (micahriggan) [inline packages/bitcore-wallet-service/src/lib/server.ts]: Missing call to next()

File: packages/bitcore-wallet-service/src/lib/chain/bch/index.ts [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+const Defaults = Common.Defaults;
+const Errors = require('../../errors/errordefinitions');
+
+export class BchChain extends BtcChain implements IChain {
+  constructor() {
+    super(BitcoreLibCash);
+  }
+
+  storeAndNotifyTx(txp, opts, cb) {
+    log.debug('Rechecking UTXOs availability for publishTx');
```
Comment (micahriggan) [inline packages/bitcore-wallet-service/src/lib/chain/bch/index.ts]: is this supposed to be `this.walletService.logd`?

File: packages/bitcore-wallet-service/src/lib/server.ts [hunk 5]
```diff
@@ -2879,52 +2637,16 @@ export class WalletService {
...
-                  if (! wallet.isUTXOCoin()) {
-                    return next();
-                  }
-                  getChangeAddress(wallet, (err, address, isNew) => {
-                    if (err) return next(err);
-                    changeAddress = address;
-
-                    return next();
-                  });
+                  changeAddress = await ChainService.getChangeAddress(wallet, opts);
```
Comment (micahriggan) [inline packages/bitcore-wallet-service/src/lib/server.ts]: try catch with next(error)

File: packages/bitcore-wallet-service/src/lib/server.ts (hunk 3 above)
Comment (micahriggan) [reply packages/bitcore-wallet-service/src/lib/server.ts]: try catch with next(err)

File: packages/bitcore-wallet-service/src/lib/server.ts [hunk 6]
```diff
@@ -2962,32 +2684,11 @@ export class WalletService {
...
-                (next) => {
-                  if (wallet.isUTXOCoin() ) {
-                    return next();
-                  }
-                  this._getTransactionCount(wallet, txp.from, (err, nonce) => {
-                    txp.nonce = nonce;
-                    return next();
-                  });
+                async(next) => {
+                  txp.nonce = await ChainService.getTransactionCount(wallet, txp.from);
```
Comment (micahriggan) [inline packages/bitcore-wallet-service/src/lib/server.ts]: try catch with next(err)

File: packages/bitcore-wallet-service/src/lib/chain/btc/index.ts [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+        const txpInputs = _.map(txp.inputs, utxoKey);
+        const utxosIndex = _.keyBy(utxos, utxoKey);
+        const unavailable = _.some(txpInputs, i => {
+          const utxo = utxosIndex[i];
+          return !utxo || utxo.locked;
+        });
+
+        if (unavailable) return cb(Errors.UNAVAILABLE_UTXOS);
+
+        txp.status = 'pending';
```
Comment (matiu) [inline packages/bitcore-wallet-service/src/lib/chain/btc/index.ts]: this should be outside the function.

File: packages/bitcore-wallet-service/src/lib/chain/bch/index.ts [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+        const txpInputs = _.map(txp.inputs, utxoKey);
+        const utxosIndex = _.keyBy(utxos, utxoKey);
+        const unavailable = _.some(txpInputs, i => {
+          const utxo = utxosIndex[i];
+          return !utxo || utxo.locked;
+        });
+
+        if (unavailable) return cb(Errors.UNAVAILABLE_UTXOS);
+
+        txp.status = 'pending';
```
Comment (matiu) [inline packages/bitcore-wallet-service/src/lib/chain/bch/index.ts]: this should be outside the function.

File: packages/bitcore-wallet-service/src/lib/chain/eth/index.ts [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+    try {
+      txp.getBitcoreTx();
+    } catch (ex) {
+      server.logw('Error building Bitcore transaction', ex);
+      return ex;
+    }
+  }
+
+  storeAndNotifyTx(server, txp, opts, cb) {
+    txp.status = 'pending';
```
Comment (matiu) [inline packages/bitcore-wallet-service/src/lib/chain/eth/index.ts]: this should be outside the function.

File: packages/bitcore-wallet-service/src/lib/chain/bch/index.ts [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+const Common = require('../../common');
+const Defaults = Common.Defaults;
+const Errors = require('../../errors/errordefinitions');
+
+export class BchChain extends BtcChain implements IChain {
+  constructor() {
+    super(BitcoreLibCash);
+  }
+
+  storeAndNotifyTx(server, txp, opts, cb) {
```
Comment (matiu) [inline packages/bitcore-wallet-service/src/lib/chain/bch/index.ts]: name => "checkTxUTXOs"?

**Comment by matiu** [comment]
great work!

**Pull request merged**